samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -461,8 +482,25 @@ public void onJobModelExpired() {
             } else {
               LOGGER.info("Container: {} shutdown completed for stream processor: {}.", container, processorId);
             }
+          } else if (state == State.IN_REBALANCE) {
+            if (container != null) {
+              boolean hasContainerShutdown = interruptContainerAndShutdownExecutorService();
+              if (!hasContainerShutdown) {
+                LOGGER.warn("Job model expire unsuccessful. Failed to interrupt container: {} safely. "
+                    + "Stopping the stream processor: {}", container, processorId);
+                state = State.STOPPING;
+                jobCoordinator.stop();
```

Comment:
It will only ensure stream processor is shutdown and propagate the application status as `UnsuccessfulFinish`. However the parent application can either choose to bring down the JVM vs continue as is. In the latter, even though we interrupted the container thread, it is possible for container components to linger especially if pluggable components use non-daemon threads.